xgi/classes/dihypergraph.py

```diff
@@ -0,0 +1,823 @@
+"""Base class for directed hypergraphs."""
+from collections.abc import Hashable, Iterable
+from copy import copy, deepcopy
+from itertools import count
+from warnings import warn
+
+from ..exception import XGIError
+from ..utils import IDDict, update_uid_counter
+from .reportviews import DiEdgeView, DiNodeView
+
+__all__ = ["DiHypergraph"]
+
+
+class DiHypergraph:
+    r"""A directed hypergraph (dihypergraph) is a collection of directed
+    interactions where each interaction has both senders and receivers.
+
+    More formally, a dihypergraph is a pair :math:`(V, E)`, where :math:`V` is
+    a set of elements called *nodes* or *vertices*, and :math:`E` is the set of
+    directed hyperedges. A directed hyperedge is an ordered pair, $(e^+, e^-)$,
+    where $e^+ \subset V$, the set of senders, is known as the "tail" and $e^-\subset V$,
+    the set of receivers, is known as the "head". The equivalent undirected edge,
+    is $e = e^+ \cap e^-$ and the edge size is defined as $|e|$.
+
+    The DiHypergraph class allows any hashable object as a node and can associate
+    attributes to each node, edge, or the hypergraph itself, in the form of key/value
+    pairs.
+
+    Multiedges and self-loops are allowed.
+
+    Parameters
+    ----------
+    incoming_data : input directed hypergraph data (optional, default: None)
+        Data to initialize the dihypergraph. If None (default), an empty
+        hypergraph is created, i.e. one with no nodes or edges.
+        The data can be in the following formats:
+
+        * directed hyperedge list
+        * directed hyperedge dictionary
+        * DiHypergraph object.
+
+    **attr : dict, optional, default: None
+        Attributes to add to the hypergraph as key, value pairs.
+
+    Notes
+    -----
+    Unique IDs are assigned to each node and edge internally and are used to refer to
+    them throughout.
+
+    The `attr` keyword arguments are added as hypergraph attributes. To add node or edge
+    attributes see :meth:`add_node` and :meth:`add_edge`.
+
+    In addition to the methods listed in this page, other methods defined in the `stats`
+    package are also accessible via the `DiHypergraph` class.  For more details, see the
+    `tutorial
+    <https://github.com/ComplexGroupInteractions/xgi/blob/main/tutorials/Tutorial%206%20-%20Statistics.ipynb>`_.
+
+    Examples
+    --------
+    >>> import xgi
+    >>> H = xgi.DiHypergraph([([1, 2, 3], [4]), ([5, 6], [6, 7, 8])])
+    >>> H.nodes
+    DiNodeView((1, 2, 3, 4, 5, 6, 7, 8))
+    >>> H.edges
+    DiEdgeView((0, 1))
```

Comment:
add example showing structure of `members()` or some output where senders and receivers are separated?

Reply:
Good idea. I did this, but because the output is a set, the order is not guaranteed, so I had to sort and convert to a list in order to pass the doctest consistently. :(